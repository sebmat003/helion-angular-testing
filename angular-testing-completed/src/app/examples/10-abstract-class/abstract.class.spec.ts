import { Component, Input } from '@angular/core';
import { FormControl, ReactiveFormsModule } from '@angular/forms';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { AbstractFormControlComponent } from './abstract.class';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';

@Component({
  selector: 'app-test',
  standalone: true,
  imports: [ReactiveFormsModule, MatFormFieldModule, MatInputModule],
  template: `
    <mat-form-field>
      <input
        matInput
        type="text"
        [formControl]="control"
        [placeholder]="placeholder"
      />
    </mat-form-field>
  `,
})
export class TestComponent extends AbstractFormControlComponent<string> {
  @Input() override placeholder = 'Type a text';
}

describe('AbstractFormControlComponent', () => {
  let spectator: Spectator<TestComponent>;
  const createComponent = createComponentFactory({
    component: TestComponent,
    imports: [ReactiveFormsModule, MatFormFieldModule, MatInputModule],
    detectChanges: false,
    declareComponent: false,
  });

  beforeEach(() => {
    spectator = createComponent();
    spectator.fixture.componentInstance.ngControl = {
      control: new FormControl(),
      valueAccessor: {},
    } as any;
  });

  it('should match the snapshot', () => {
    spectator.detectChanges();

    expect(spectator.fixture).toMatchSnapshot();
  });

  it('should initialize with the correct placeholder', () => {
    spectator.detectChanges();

    expect((spectator.query('input') as HTMLInputElement).placeholder).toBe(
      'Type a text'
    );
  });

  it('should update value on input', () => {
    spectator.detectChanges();
    const input = spectator.query('input') as HTMLInputElement;

    input.value = 'test';
    input.dispatchEvent(new Event('input'));
    spectator.detectChanges();

    expect(spectator.component.control.value).toBe('test');
  });
});
