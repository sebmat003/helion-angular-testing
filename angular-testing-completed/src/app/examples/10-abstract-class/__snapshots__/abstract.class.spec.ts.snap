// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbstractFormControlComponent should match the snapshot 1`] = `
<app-test
  ngControl={[Function Object]}
  placeholder={[Function String]}
>
  <mat-form-field
    class="mat-mdc-form-field mat-mdc-form-field-type-mat-input mat-form-field-no-animations mat-form-field-appearance-fill mat-primary ng-untouched ng-pristine ng-valid"
  >
    <div
      class="mat-mdc-text-field-wrapper mdc-text-field mdc-text-field--filled mdc-text-field--no-label"
    >
      <div
        class="mat-mdc-form-field-focus-overlay"
      />
      <div
        class="mat-mdc-form-field-flex"
      >
        <div
          class="mat-mdc-form-field-infix"
        >
          <input
            aria-invalid="false"
            aria-required="false"
            class="mat-mdc-input-element ng-untouched ng-pristine ng-valid mat-mdc-form-field-input-control mdc-text-field__input cdk-text-field-autofill-monitored"
            matinput=""
            placeholder="Type a text"
            type="text"
          />
        </div>
      </div>
      <div
        class="mdc-line-ripple mdc-line-ripple--deactivating"
        matformfieldlineripple=""
      />
    </div>
    <div
      class="mat-mdc-form-field-subscript-wrapper mat-mdc-form-field-bottom-align"
    >
      <div
        class="mat-mdc-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
        style="opacity: 1; transform: translateY(0%);"
      >
        <div
          class="mat-mdc-form-field-hint-spacer"
        />
      </div>
    </div>
  </mat-form-field>
</app-test>
`;
